### ISSUE TITLE
IsolatedProcess :  No implementation found for void mono.android.Runtime.register

### ISSUE THREAD

**Issue by calraiden** [xamarin/xamarin-android#3378]
### Steps to Reproduce

1.  Create Custom Service Class with IsolatedProcess = true
2.  Call Service in the MainActivity
3.  The isolated process init, but show error.

I develop an Xamarin.Droid App and need created an isolated process. But, I received error when call de intent service in the MainActivity.

```
[Activity(Label = "@string/app_title", Theme = "@style/AppTheme.NoActionBarNotExtended", MainLauncher = false, ScreenOrientation = Android.Content.PM.ScreenOrientation.Portrait, WindowSoftInputMode = SoftInput.StateAlwaysHidden | SoftInput.AdjustPan)]
    public class MainActivity : AppCompatActivity, NavigationView.IOnNavigationItemSelectedListener, FloatingActionButton.IOnClickListener
    {
		...
        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
			...
			//Isolated
            StartService(new Intent(this, typeof(Services.Periodic.IsolatedService)));
		}			
	}
}
```

it fails to start the Service process and crashes stating it cannot resolve overloads from Java.

```
using Android.App;
using Android.Content;
using Android.OS;
using Android.Util;
using System;

namespace Ishpia.Droid.Activities.Services.Periodic
{
    [Service(Exported = true, 
        Enabled = true, 
        IsolatedProcess = true,
        Process = "com.bergmannsoft.aprepara.isolatedservice.isolated_process")]
    class IsolatedService : Service
    {
        static readonly string TAG = typeof(IsolatedService).FullName;

        public IBinder Binder { get; private set; }

        public override void OnCreate()
        {
            // This method is optional to implement
            base.OnCreate();
            Log.Debug(TAG, "IsolatedService OnCreate");
        }

        public override IBinder OnBind(Intent intent)
        {
            // This method must always be implemented
            Log.Debug(TAG, "IsolatedService OnBind");
            return this.Binder;
        }

        public override bool OnUnbind(Intent intent)
        {
            // This method is optional to implement
            Log.Debug(TAG, "IsolatedService OnUnbind");
            return base.OnUnbind(intent);
        }

        public override void OnDestroy()
        {
            // This method is optional to implement
            Log.Debug(TAG, "IsolatedService OnDestroy");
            Binder = null;
            base.OnDestroy();
        }
    }
}
```
Error:
```
07-22 12:15:01.742: E/AndroidRuntime(5767): FATAL EXCEPTION: main
07-22 12:15:01.742: E/AndroidRuntime(5767): Process: com.bergmannsoft.aprepara.isolatedservice.isolated_process, PID: 5767
07-22 12:15:01.742: E/AndroidRuntime(5767): java.lang.UnsatisfiedLinkError: No implementation found for void mono.android.Runtime.register(java.lang.String, java.lang.Class, java.lang.String) (tried Java_mono_android_Runtime_register and Java_mono_android_Runtime_register__Ljava_lang_String_2Ljava_lang_Class_2Ljava_lang_String_2)
07-22 12:15:01.742: E/AndroidRuntime(5767): 	at mono.android.Runtime.register(Native Method)
07-22 12:15:01.742: E/AndroidRuntime(5767): 	at md527cd63edec29ae2b7f288c3ee88495be.IsolatedService.<clinit>(IsolatedService.java:18)
07-22 12:15:01.742: E/AndroidRuntime(5767): 	at java.lang.Class.newInstance(Native Method)
07-22 12:15:01.742: E/AndroidRuntime(5767): 	at android.app.ActivityThread.handleCreateService(ActivityThread.java:3321)
07-22 12:15:01.742: E/AndroidRuntime(5767): 	at android.app.ActivityThread.-wrap4(Unknown Source:0)
07-22 12:15:01.742: E/AndroidRuntime(5767): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1677)
07-22 12:15:01.742: E/AndroidRuntime(5767): 	at android.os.Handler.dispatchMessage(Handler.java:106)
07-22 12:15:01.742: E/AndroidRuntime(5767): 	at android.os.Looper.loop(Looper.java:164)
07-22 12:15:01.742: E/AndroidRuntime(5767): 	at android.app.ActivityThread.main(ActivityThread.java:6494)
07-22 12:15:01.742: E/AndroidRuntime(5767): 	at java.lang.reflect.Method.invoke(Native Method)
07-22 12:15:01.742: E/AndroidRuntime(5767): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
07-22 12:15:01.742: E/AndroidRuntime(5767): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)

```
I found some comments about the problem, but I did not get it resolved.
https://xamarin.github.io/bugzilla-archives/51/51940/bug.html#

Please, can anyone help me?

**Comment by calraiden** [comment]
![image](https://user-images.githubusercontent.com/1984844/61632610-a4e93f00-ac63-11e9-8b3b-3681cbf2c1c1.png)

**Comment by dellis1972** [comment]
@calraiden can you post the full output of the `adb logcat`, it might contain some additional information that we need. I suspect what is happening is that the native libraries are not being resolved in this case.

**Comment by calraiden** [comment]
> @calraiden can you post the full output of the `adb logcat`, it might contain some additional information that we need. I suspect what is happening is that the native libraries are not being resolved in this case.

Hi @dellis1972 ,

I'm attached the log of visual studio and DDMS.

Do you Need more data?

Thank you in advance =)


[logvs.txt](https://github.com/xamarin/xamarin-android/files/3422452/logvs.txt)
[logcat.txt](https://github.com/xamarin/xamarin-android/files/3422453/logcat.txt)

**Comment by grendello** [comment]
Hi @calraiden! While I don't have a solution for you yet, I can offer a tentative explanation of what's happening. The cause is simple - isolated service is started in a new process in which our runtime is **not** initialized by Android and, hence, our Java methods implemented in C++ are not registered with the Java VM and so it fails to find the method you see in the error message. As for the solution,I don't have one yet, albeit it appears "simple" - initialize Xamarin.Android runtime within the new process.

**Comment by calraiden** [comment]
Hi @grendello ,

As a programmer, I understand that the simplest bugs are the most complex (sometimes ).

Thanks for the feedback.

**Comment by grendello** [comment]
Hi @calraiden, thanks for your understanding, I know how frustrating it is to be blocked on some issue external to your app! Alas, my investigation so far doesn't give hope for a straightforward solution to your problem :( What happens behind the scenes when you start an isolated service is that Android forks a new process in the **system** context, without any rights to access any files etc of the parent app - the only way you can communicate with the new service is via the Service APIs. Unfortunately, that means we aren't able to load our runtimes (XA and Mono). Here's a log excerpt from Android Q:

```
08-21 12:30:50.609 16656 16656 I imestampServic: The ClassLoaderContext is a special shared library.
08-21 12:30:50.610 16656 16656 W imestampService: type=1400 audit(0.0:1871): avc: denied { getattr } for path="/data/data/com.xamarin.xample.startedservicesdemo" dev="dm-6" ino=17061 scontext=u:r:isolated_app:s0:c40,c291,c512,c768 tcontext=u:object_r:app_data_file:s0:c512,c768 tclass=dir permissive=0
08-21 12:30:50.620 16656 16656 I chatty  : uid=99000(u0_i9000) com.xamarin.xample.startedservicesdemo:md52a460aa7510222321abfcee40d17b122.TimestampService identical 2 lines
08-21 12:30:50.620 16656 16656 W imestampService: type=1400 audit(0.0:1874): avc: denied { getattr } for path="/data/data/com.xamarin.xample.startedservicesdemo" dev="dm-6" ino=17061 scontext=u:r:isolated_app:s0:c40,c291,c512,c768 tcontext=u:object_r:app_data_file:s0:c512,c768 tclass=dir permissive=0
08-21 12:30:50.631 16656 16656 E imestampServic: No implementation found for void mono.android.Runtime.register(java.lang.String, java.lang.Class, java.lang.String) (tried Java_mono_android_Runtime_register and Java_mono_android_Runtime_register__Ljava_lang_String_2Ljava_lang_Class_2Ljava_lang_String_2)
08-21 12:30:50.632 16656 16656 D AndroidRuntime: Shutting down VM
```

Especially note the audit warnings on lines 2 and 4 - this is what prevents us from loading our DSOs. For now, I don't see any other option than to implement your isolated service in Java :( I'll keep digging, but I wouldn't rise your hopes up too high.

**Comment by fernandovm** [comment]
Dear, some news about this issue? Since upgrade to xamarin forms 4 I receiving app crash reports of this issue on the Google Play Console, but I not receiving any report on the appcenter. I'm not using an IsolatedProcess service. 

java.lang.UnsatisfiedLinkError: 
  at mono.android.Runtime.register (Native Method)
  at studying.mainService.<clinit>
  at java.lang.Class.newInstance (Native Method)
  at android.app.ActivityThread.handleCreateService (ActivityThread.java:3528)
  at android.app.ActivityThread.-wrap6 (ActivityThread.java)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1749)
  at android.os.Handler.dispatchMessage (Handler.java:102)
  at android.os.Looper.loop (Looper.java:154)
  at android.app.ActivityThread.main (ActivityThread.java:6823)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:1563)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1451)


![image](https://user-images.githubusercontent.com/441113/78950872-869fc980-7aa6-11ea-8aad-61ba36595e6f.png)

![image](https://user-images.githubusercontent.com/441113/78950881-8b647d80-7aa6-11ea-9a3c-860081d41388.png)

![image](https://user-images.githubusercontent.com/441113/78950884-8ef80480-7aa6-11ea-8e31-8879bb70b8ef.png)

I don't receive any additional information about it, how I can deal with this issue? Any tip? I'm already update xamarin forms to 4.5.0.530, but the problem remains.

**Comment by brendanzagaeski** [comment]
> I'm not using an IsolatedProcess service.

@fernandovm, if you're not using an IsolatedProcess, then the issue you're seeing is likely separate from this issue #3378. The best next step would be to submit a [new issue](https://github.com/xamarin/xamarin-android/issues/new) so the team can take a look at your specific issue. If possible, attach a sample project that demonstrates the issue or the `adb logcat` output from your app. Thanks in advance!

**Comment by Samaun91** [comment]
Any news on this topic?

**Comment by grendello** [comment]
@Samaun91 alas, no. As stated in the [comment above](https://github.com/xamarin/xamarin-android/issues/3378#issuecomment-523401992), it's nearly impossible to implement support for this, at least with the way Xamarin.Android currently works.  I have one idea we could try, but it would require some serious changes to the product architecture, something I don't think can be done now.

**Comment by jonathanpeppers** [comment]
It looks like this problem also happens via `android:process`, so an example is the Baidu Push SDK:

https://github.com/dotnet/maui/issues/16142

**Comment by busec0** [comment]
@jonathanpeppers The problem with binding `Baidu Push SDK` is a bit different, but the root cause is the same. You have the separate process written in Java to listen for push notifications. It tries to use the `ApplicationContext`'s `OnCreate`. Now, you have two cases: 

1. You override the `OnCreate` on Xamarin -> **Failure** 
2. You **DO NOT** override it -> **It works.** 

### This is what happens in Java bytecode when you override: 
![Screenshot 2023-07-19 at 10 56 52](https://github.com/xamarin/xamarin-android/assets/26183235/39ce1953-ed6d-455c-872d-cfbfd87b668f)

This is the bytecode for `OnCreate`, which calls `n_onCreate`. I'd assume this is where the Xamarin bridge happens to reach to C#. 
```
.method public onCreate()V
    .registers 1

    .line 25
    invoke-direct {p0}, Lcrc64c0e3dc4476436a10/MyApplication;->n_onCreate()V

    .line 26
    return-void
.end method
```

### If you do not override: 

`OnCreate` is of course not there. There's no bridge to Xamarin. It works fine. 

![Screenshot 2023-07-19 at 11 02 18](https://github.com/xamarin/xamarin-android/assets/26183235/4643222e-0222-458a-a2f8-0e0ceec66e7d)


### Conclusion 

For Xamarin Android Native this has workarounds like simply avoiding to override `OnCreate` or use `new` and call base. For MAUI apps, you can't, because the OnCreate override is bundled in `MauiApplication` -> you'll always end up with the Xamarin bridge in bytecode. 

Now, my question would be, given this is unlikely to be fix: 

1. Given these extra details, can you think of a way to work around it? 
2. Is there a way to `try-catch` the `n_onCreate` call? You won't get the C# code executed, but at least the secondary process will not be killed.

**Comment by neobenedict** [comment]
This prevents use of shizuku with Xamarin.

Shame microsoft has killed xamarin, this will never be fixed.

**Comment by jonathanpeppers** [comment]
> microsoft has killed xamarin, this will never be fixed.

I'm confused by this comment, as the Android workload for .NET shipped in .NET 6, 7, 8 and we are currently working on .NET 9. We ship the Android workload out of this repo.

Support policy:

https://dotnet.microsoft.com/en-us/platform/support/policy/maui

**Comment by neobenedict** [comment]
Just renamed/merged then, that's good news.

**Comment by Evengard** [comment]
Stumbled upon that problem...
Honestly, it looks to me like there's just a need for a slim java wrapper around mono.android.Runtime.register which checks if Mono is initialized and initialize it if it isn't. No direct jump to the native code, but a wrapper instead. Looks like there's a need to update the codegen for that, but that seems to not be complex. Too bad I'm not familiar with xamarin codebase, I'd love to do a PR for that but I'm afraid I'm gonna be lost when searching for where that should be done.

**Comment by dellis1972** [comment]
> Stumbled upon that problem... Honestly, it looks to me like there's just a need for a slim java wrapper around mono.android.Runtime.register which checks if Mono is initialized and initialize it if it isn't. No direct jump to the native code, but a wrapper instead. Looks like there's a need to update the codegen for that, but that seems to not be complex. Too bad I'm not familiar with xamarin codebase, I'd love to do a PR for that but I'm afraid I'm gonna be lost when searching for where that should be done.

Maybe you can test out that theory using the `AndroidJavaSource` build action in a test app?

**Comment by Evengard** [comment]
I actually digged a bit and stumbled upon a somewhat block. For initializing Xamarin I need a context, to retrieve all sort of information about the app. There aren't a lot of ways getting that one which are "legal" (eg not using some internal android apis - which is possible but is frowned upon).

I'm thinking now about trying to "defer" them if it is not initialized and then just call em during the initialization. Seems easier and more logical...

**Comment by redd-tsu** [comment]
Seeing the same error trying to implement `AppComponentFactory`. 

From what I gather here the component factory runs too early? Interested if there are any workarounds (perhaps to force Xamarin to init earlier?).

**Comment by Evengard** [comment]
As far as I can see, Xamarin is basically loaded as a component, and in some cases the component isn't loaded, hence Xamarin isn't loaded too, or something like that

**Comment by alex-relov** [comment]
Hello, is there any news about IsolatedProcess?
